File: src/daegun/week25/시소 짝꿍.swift
/*
 시간 초과
 */

import Foundation

func solution(_ weights:[Int]) -> Int64 {
    
    let distances = [2, 3, 4]
    
    var possibleWeight = [Int:Set<Int>]()
    
    var result: Int64 = 0
    for (index, weight) in weights.enumerated() {
        var pair = Set<Int>()
        for distance in distances {
            
            let multiplied = distance * weight
            
            if let existingSet = possibleWeight[multiplied] {
                var newSet = existingSet
                newSet.insert(index)
                possibleWeight[multiplied] = newSet
                
                for item in newSet where item != index {
                    pair.insert(item)
                }
                
            } else {
                possibleWeight[multiplied] = [index]
            }
        }
        result += Int64(pair.count)
    }
    
    return result
}

File: src/taeeun/week25/시소 짝꿍.cpp
#include <string>
#include <vector>
#include <map>
#include <set>

using namespace std;

long long solution(vector<int> weights) {
    long long answer = 0;
    vector<int> d={2, 3, 4};
    vector<int> n={1, 2, 3};
    set<int> s;
    map<long, long> m;

    for (int i=0; i<weights.size(); i++) {
        s.insert(weights[i]);
        m[weights[i]]+=1;
    }

    for (auto weight: s) {
        //1:1 비율
        answer+=(m[weight]*(m[weight]-1))/2;

        //그외 비율
        for (int i=0; i<3; i++) {
            if (weight%n[i]) {
                continue;
            }
            answer+=m[(weight/n[i])*d[i]]*m[weight];
        }
    }

    return answer;
}